sstables/checksummed_data_source.cc

```diff
@@ -27,16 +28,26 @@ template <ChecksumUtils ChecksumType>
 class checksummed_file_data_source_impl : public data_source_impl {
     std::optional<input_stream<char>> _input_stream;
     const checksum& _checksum;
+    std::optional<const uint32_t> _expected_digest;
+    uint32_t _actual_digest;
+    digest_validation_result* _digest_result;
     uint64_t _chunk_size_trailing_zeros;
+    uint64_t _file_len;
     uint64_t _underlying_pos;
     uint64_t _pos;
     uint64_t _beg_pos;
     uint64_t _end_pos;
 public:
     checksummed_file_data_source_impl(file f, uint64_t file_len,
                 const checksum& checksum, uint64_t pos, size_t len,
-                file_input_stream_options options)
+                file_input_stream_options options,
+                std::optional<uint32_t> digest,
+                digest_validation_result* digest_result)
```

Comment:
Since you throw on all cases where the digest validation fails, I think this output param can be just dropped.